src/java/org/apache/cassandra/locator/ReplicaList.java

```diff
@@ -0,0 +1,244 @@
+    {
+        this.replicaList = replicaList;
+    }
+
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ReplicaList that = (ReplicaList) o;
+        return Objects.equals(replicaList, that.replicaList);
+    }
+
+    public int hashCode()
+    {
+
+        return Objects.hash(replicaList);
+    }
+
+    @Override
+    public String toString()
+    {
+        return replicaList.toString();
+    }
+
+    @Override
+    public boolean add(Replica replica)
+    {
+        return replicaList.add(replica);
+    }
+
+    @Override
+    public void addAll(Iterable<Replica> replicas)
+    {
+        Iterables.addAll(replicaList, replicas);
+    }
+
+    @Override
+    public int size()
+    {
+        return replicaList.size();
+    }
+
+    @Override
+    public Iterator<Replica> iterator()
+    {
+        return replicaList.iterator();
+    }
+
+    public Replica get(int idx)
+    {
+        return replicaList.get(idx);
+    }
+
+    public List<InetAddressAndPort> asEndpointList()
+    {
+        List<InetAddressAndPort> endpoints = new ArrayList<>(replicaList.size());
+        for (Replica replica: replicaList)
+        {
+            endpoints.add(replica.getEndpoint());
+        }
+        return endpoints;
+    }
+
+    @Override
+    public void removeEndpoint(InetAddressAndPort endpoint)
+    {
+        replicaList.removeIf(r -> r.getEndpoint().equals(endpoint));
+    }
+
+    @Override
+    public void removeReplica(Replica replica)
+    {
+        replicaList.remove(replica);
+    }
+
+    public ReplicaList filter(Predicate<Replica> predicate)
+    {
+        ArrayList<Replica> newReplicaList = new ArrayList<>(size());
+        for (Replica replica: replicaList)
+        {
+            if (predicate.test(replica))
+            {
+                newReplicaList.add(replica);
+            }
+        }
+        return new ReplicaList(newReplicaList);
+    }
+
+    public void sort(Comparator<Replica> comparator)
+    {
+        replicaList.sort(comparator);
+    }
+
+    public static ReplicaList intersectEndpoints(ReplicaList l1, ReplicaList l2)
+    {
+        Replicas.checkFull(l1);
+        Replicas.checkFull(l2);
+        // Note: we don't use Guava Sets.intersection() for 3 reasons:
+        //   1) retainAll would be inefficient if l1 and l2 are large but in practice both are the replicas for a range and
+        //   so will be very small (< RF). In that case, retainAll is in fact more efficient.
+        //   2) we do ultimately need a list so converting everything to sets don't make sense
+        //   3) l1 and l2 are sorted by proximity. The use of retainAll  maintain that sorting in the result, while using sets wouldn't.
+        Collection<InetAddressAndPort> endpoints = l2.asEndpointList();
+        return l1.filter(r -> endpoints.contains(r.getEndpoint()));
+    }
+
+    public static ReplicaList of(Replica... replicas)
```

Comment:
fixed

Reply:
Hmmm so what I thought was a specialization would be possible, but since it's mutable you have to allocate a real list. So really I think it's fine to have of() allocate the wrapper array. So of() could just return a global immutable one. The singleton one could do a singleton list wrapper.

TL;DR this is not important enough to use up more of your time. You can delete the specializations.